keras/backend/theano_backend.py

```diff
@@ -501,6 +501,49 @@ def function(inputs, outputs, updates=[], **kwargs):
 def gradients(loss, variables):
     return T.grad(loss, variables)
 
+# SHAPE INFERENCE
+
+def infer_shape(func, input_shape):
+    '''Automatic shape inference
+
+    # Arguments
+        func: Function, lambda or callable object which accepts a tensor or a list of tensors
+        and returns a tensor or a list of tensors.
+        input_shape: tuple or list of tuples. Shape(s) of input(s).
+
+    # Returns
+        tuple if func returns a single tensor, list of tuple if func returns a list of tensors.
+    '''
+    if not type(input_shape[0]) in [list, tuple]:
+        input_shapes = [input_shape]
+    else:
+        input_shapes = input_shape
+    input_shapes_1 = list(map(list, input_shapes))
+    input_shapes_2 = list(map(list, input_shapes))
+    for i in range(len(input_shapes)):
+        for j in range(len(input_shapes[i])):
+            if not input_shapes_1[i][j]:
+                input_shapes_1[i][j] = 10
+                input_shapes_2[i][j] = 0
```

Comment:
If we set it to, say 20, then x[:5] would give the same shape for both input shapes, and will not be set to None.